### PR TITLE
add python 3.10 support

### PR DESCRIPTION
**Pull request by ebattat**
This PR include:

1. Dockerfile: Add Python 3.10 support
2. Dockerfile: Update oc client version to 4.11.0-0.okd-2022-10-15-073651
3. Update Python 3.10 in setup.py
4. GitHub Actions: Add unittest for python 3.6, 3.7, 3.8, 3.9, 3.10
5. GitHub Actions: Update integration test for Python 3.10
6. GitHub Actions: Update oc client version to 4.11.0-0.okd-2022-10-15-073651
7. Add is_pod_exist_by_label method

### CODE REVIEW

File: .github/workflows/Func_Env_Build_Test_CI.yml [hunk 1]
```diff
@@ -14,19 +14,85 @@ concurrency:
...
+          # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+          flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+      - name: ⚙️ Set Kubeconfig and hosts
+        env:
+          KUBECONFIG: ${{ secrets.FUNC_KUBECONFIG }}
+          RUNNER_PATH: ${{ secrets.RUNNER_PATH }}
+          OCP_HOSTS: ${{ secrets.FUNC_OCP_HOSTS }}
+        run: |
+          mkdir -p "$RUNNER_PATH/.kube/"
+          echo "$KUBECONFIG" > "$RUNNER_PATH/.kube/config"
```
Comment (RobertKrawitz) [inline .github/workflows/Func_Env_Build_Test_CI.yml:50]: ```cp "$KUBECONFIG" "$RUNNER_PATH/.kube/config"```

File: .github/workflows/Func_Env_Build_Test_CI.yml [hunk 2]
```diff
@@ -14,19 +14,85 @@ concurrency:
...
+          IBM_SECRET_ACCESS_KEY: ${{ secrets.IBM_SECRET_ACCESS_KEY }}
+          IBM_BUCKET: ${{ secrets.IBM_BUCKET }}
+          IBM_KEY: ${{ secrets.IBM_KEY }}
+          RUN_ARTIFACTS_URL: ${{ secrets.FUNC_RUN_ARTIFACTS_URL }}
+        run: |
+          # Install Dockerfile content for pytest
+          # install oc/kubectl
+          oc_version=4.11.0-0.okd-2022-10-15-073651
+          curl -L https://github.com/openshift/okd/releases/download/${oc_version}/openshift-client-linux-${oc_version}.tar.gz -o $RUNNER_PATH/openshift-client-linux-${oc_version}.tar.gz
+          tar -xzvf $RUNNER_PATH/openshift-client-linux-${oc_version}.tar.gz -C $RUNNER_PATH/
```
Comment (RobertKrawitz) [inline .github/workflows/Func_Env_Build_Test_CI.yml:75]: Don't need the temporary file:

```curl -L "https://github.com/openshift/okd/releases/download/${oc_version}/openshift-client-linux-${oc_version}.tar.gz" | tar xvf - -C "$RUNNER_PATH/"```

File: .github/workflows/Func_Env_Build_Test_CI.yml [hunk 3]
```diff
@@ -14,19 +14,85 @@ concurrency:
...
+  integration_test:
+    name: integration_test
+    needs: [unittest]
     runs-on: ubuntu-latest
-    outputs:
-       start_time_output: ${{ steps.start_step.outputs.start_time }}
     steps:
     - uses: actions/checkout@v3
-    - id: start_step
-      run: echo "start_time=$(printf '%(%s)T' -1)" >> $GITHUB_OUTPUT
```
Comment (RobertKrawitz) [inline .github/workflows/Func_Env_Build_Test_CI.yml:25]: Why are you removing this?

File: .github/workflows/Func_Env_PR_Test_CI.yml [hunk 4]
```diff
@@ -15,17 +15,88 @@ concurrency:
...
+          # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+          flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+      - name: ⚙️ Set Kubeconfig and hosts
+        env:
+          KUBECONFIG: ${{ secrets.FUNC_KUBECONFIG }}
+          RUNNER_PATH: ${{ secrets.RUNNER_PATH }}
+          OCP_HOSTS: ${{ secrets.FUNC_OCP_HOSTS }}
+        run: |
+          mkdir -p "$RUNNER_PATH/.kube/"
+          echo "$KUBECONFIG" > "$RUNNER_PATH/.kube/config"
```
Comment (RobertKrawitz) [inline .github/workflows/Func_Env_PR_Test_CI.yml:52]: See above -- copy the file, not echo the KUBECONFIG path.

File: .github/workflows/Func_Env_PR_Test_CI.yml [hunk 5]
```diff
@@ -15,17 +15,88 @@ concurrency:
...
+          IBM_SECRET_ACCESS_KEY: ${{ secrets.IBM_SECRET_ACCESS_KEY }}
+          IBM_BUCKET: ${{ secrets.IBM_BUCKET }}
+          IBM_KEY: ${{ secrets.IBM_KEY }}
+          RUN_ARTIFACTS_URL: ${{ secrets.FUNC_RUN_ARTIFACTS_URL }}
+        run: |
+          # Install Dockerfile content for pytest
+          # install oc/kubectl
+          oc_version=4.11.0-0.okd-2022-10-15-073651
+          curl -L https://github.com/openshift/okd/releases/download/${oc_version}/openshift-client-linux-${oc_version}.tar.gz -o $RUNNER_PATH/openshift-client-linux-${oc_version}.tar.gz
+          tar -xzvf $RUNNER_PATH/openshift-client-linux-${oc_version}.tar.gz -C $RUNNER_PATH/
```
Comment (RobertKrawitz) [inline .github/workflows/Func_Env_PR_Test_CI.yml:77]: See above, this doesn't need a temporary file.

File: Dockerfile [hunk 6]
```diff
@@ -14,15 +14,25 @@ Run dnf config-manager --set-enabled powertools \
...
-RUN dnf install -y python3.9 \
-    && echo alias python=python3.9 >> ~/.bashrc
+# Prerequisite for Python installation
+ARG python_version=3.10
+ARG python_full_version=3.10.8
+RUN dnf install openssl-devel bzip2-devel wget -y
+
+# Install Python
+RUN wget https://www.python.org/ftp/python/${python_full_version}/Python-${python_full_version}.tgz \
+    && tar -xzf Python-${python_full_version}.tgz \
```
Comment (RobertKrawitz) [inline Dockerfile:24]: Again, you don't need a temporary file here.

File: benchmark_runner/common/oc/oc.py [hunk 7]
```diff
@@ -227,8 +227,8 @@ def _get_pod_name(self, pod_name: str,
         except Exception as err:
             raise PodNameNotExist(pod_name=pod_name)
 
-    def _is_pod_exist(self, pod_name: str,
-                      namespace: str = environment_variables.environment_variables_dict['namespace']):
+    def is_pod_exist(self, pod_name: str,
```
Comment (RobertKrawitz) [inline benchmark_runner/common/oc/oc.py:230]: If you're going to change the name, make it correct: `does_pod_exist` or `pod_exists`

File: benchmark_runner/common/oc/oc.py [hunk 8]
```diff
@@ -241,10 +241,26 @@ def _is_pod_exist(self, pod_name: str,
         else:
             return False
 
+    def is_pod_exist_by_label(self, label_name: str,
```
Comment (RobertKrawitz) [inline benchmark_runner/common/oc/oc.py:244]: `does_pod_exist_by_label`

File: benchmark_runner/common/oc/oc.py [hunk 9]
```diff
@@ -253,9 +269,11 @@ def _get_vm_name(self, vm_name: str, namespace: str):
             raise VMNameNotExist(vm_name=vm_name)
 
     @typechecked
-    def _is_vm_exist(self, vm_name: str, namespace: str):
+    def is_vm_exist(self, vm_name: str, namespace: str):
```
Comment (RobertKrawitz) [inline benchmark_runner/common/oc/oc.py:272]: `does_vm_exist`

**Review by RobertKrawitz — CHANGES_REQUESTED** [review]

File: .github/workflows/Func_Env_Build_Test_CI.yml (hunk 1 above)
Comment (ebattat) [reply .github/workflows/Func_Env_Build_Test_CI.yml:50]: "$KUBECONFIG" is not a file, it a string

File: .github/workflows/Func_Env_Build_Test_CI.yml (hunk 2 above)
Comment (ebattat) [reply .github/workflows/Func_Env_Build_Test_CI.yml:75]: its running inside Ubuntu container, working for me, prefer not to change it

File: .github/workflows/Func_Env_Build_Test_CI.yml (hunk 3 above)
Comment (ebattat) [reply .github/workflows/Func_Env_Build_Test_CI.yml:25]: "$KUBECONFIG" is not a file, it a string

File: .github/workflows/Func_Env_PR_Test_CI.yml (hunk 5 above)
Comment (ebattat) [reply .github/workflows/Func_Env_PR_Test_CI.yml:77]: See above replay

File: Dockerfile (hunk 6 above)
Comment (ebattat) [reply Dockerfile:24]: See above reply

File: benchmark_runner/common/oc/oc.py (hunk 7 above)
Comment (ebattat) [reply benchmark_runner/common/oc/oc.py:230]: I will change it to pod_exists

File: benchmark_runner/common/oc/oc.py (hunk 8 above)
Comment (ebattat) [reply benchmark_runner/common/oc/oc.py:244]: I will change it to: `label_pod_exists`

File: benchmark_runner/common/oc/oc.py (hunk 9 above)
Comment (ebattat) [reply benchmark_runner/common/oc/oc.py:272]: I will change it to: vm_exists

File: .github/workflows/Func_Env_Build_Test_CI.yml (hunk 1 above)
Comment (RobertKrawitz) [reply .github/workflows/Func_Env_Build_Test_CI.yml:50]: In that case, rename KUBECONFIG to something else, such as KUBECONFIG_CONTENTS.  KUBECONFIG normally refers to a filename, not the contents of the file.

File: .github/workflows/Func_Env_Build_Test_CI.yml (hunk 2 above)
Comment (RobertKrawitz) [reply .github/workflows/Func_Env_Build_Test_CI.yml:75]: This isn't that big of a deal, but it is not necessary and if the tar fails the file will be left around (and yes, the container exit will clean it up, but it's best not to leave things like that around).

Comment (ebattat) [reply .github/workflows/Func_Env_Build_Test_CI.yml:75]: Tried it on my local and it failed:
```
$ curl -L "https://github.com/openshift/okd/releases/download/4.11.0-0.okd-2022-10-15-073651/openshift-client-linux-4.11.0-0.okd-2022-10-15-073651.tar.gz" | tar xvf - -C "/tmp/"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0tar: Archive is compressed. Use -z option
tar: Error is not recoverable: exiting now
  0 50.1M    0 17471    0     0   5823      0  2:30:25  0:00:03  2:30:22 32234
curl: (23) Failed writing body (335 != 1422)
```

**Review by RobertKrawitz — APPROVED** [review]
I think you should get rid of those temporary files, but that's not enough to hold it up.  So I am approving.

**Pull request merged**